src/error.rs

```diff
@@ -139,3 +139,67 @@ where
 {
     DbErr::Json(s.to_string())
 }
+
+/// An error from unsuccessful SQL query
+#[derive(Error, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum SqlErr {
+    /// error for inserting a record with a key that already exists in the table
+    #[error("Cannot have record with same key")]
+    UniqueConstraintViolation(),
+    /// error for Foreign key is not primary key
+    #[error("Cannot add non-primary key from other table")]
+    ForeignKeyConstraintViolation(),
+}
+
+#[allow(dead_code)]
+impl DbErr {
+    /// convert generic DbErr by sqlx to SqlErr
+    /// return none if input is not any of SqlErr
+    pub fn sql_err(&self) -> Option<SqlErr> {
+        #[cfg(any(
+            feature = "sqlx-mysql",
+            feature = "sqlx-postgres",
+            feature = "sqlx-sqlite"
+        ))]
+        {
+            use std::ops::Deref;
+            if let DbErr::Exec(RuntimeErr::SqlxError(sqlx::Error::Database(e)))
+            | DbErr::Query(RuntimeErr::SqlxError(sqlx::Error::Database(e))) = self
+            {
+                let error_code = e.code().unwrap_or_default();
+                let error_code_expanded = error_code.deref();
+                #[cfg(feature = "sqlx-mysql")]
+                if e.try_downcast_ref::<sqlx::mysql::MySqlDatabaseError>()
+                    .is_some()
+                {
+                    match error_code_expanded {
+                        "23000" => return Some(SqlErr::UniqueConstraintViolation()),
+                        "1586" => return Some(SqlErr::UniqueConstraintViolation()),
```

Comment:
I wonder why we have two MySQL error codes representing the same unique constraint violation error?

Can we have a comment documenting the corresponding web page of each database's error code?

Reply:
Originally I put 23000 since I saw it in the error testing file, but later I found that 23000 is the sqlstate for both unique/foreign key constraint violation. Therefore I cannot use sqlstate there. I would update the error codes again once I make a foreign key constraint test. 
Thank you for spotting this though!